% chapter02.tex

 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %                                                                           %
 %    YABBY documentation                                                    %
 %    Copyright (C) 2007 Vladimir Likic                                      %
 %                                                                           %
 %    The files in this directory provided under the Creative Commons        %
 %    Attribution-NonCommercial-NoDerivs 2.1 Australia license               %
 %    http://creativecommons.org/licenses/by-nc-nd/2.1/au/                   %
 %    See the file license.txt                                               %
 %                                                                           %
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\setcounter{section}{0}

\chapter{Tutorial}

The data files used in this tutorial can be found in docs/data/ 
directory. Starting Yabby in this directory should allow one to
execute all examples given in the tutorial.

\section{Working with sequences}

\index{sequences}

In Yabby sequences are represented as sequence objects. A sequence
object may contain one or more sequences. 

One way to create a sequence object is to load sequences from a
file.  Consider the file 'cad3.fas' which contains three sequences
from the Pfam CAD family in the FASTA format:

\begin{verbatim}
>Q53650_STAAU
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>Q97PJ0_STRPN
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>P95773_STALU
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

\index{seq\_load}

To load this file in the workspace use the command 'seq\_load':

\begin{verbatim}
yabby> seq_load cad3.fas cad3

 Reading the file 'cad3.fas' ..
 3 sequence(s) found.
\end{verbatim}

Most Yabby commands which create a new object in the workspace
conform to the same pattern:

\begin{verbatim}
COMMAND [ options ] ARGS OBJ_NAME
\end{verbatim}

Where '[ options ]' is where flags and options are passed (if
any), ARGS are the arguments to the command, and OBJ\_NAME is
the name of the object to be created. In the example given
above, there were no options to the 'seq\_load' command; there
was one argument, the name of the file; and the object to be
created was named 'cad3'.

It is possible to inspect objects currently available in the
workspace:

\index{what}

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
\end{verbatim}

This listing shows that one object is currently in the workspace,
of the 'seq' type. It is possible to load the same sequences
under a different name: 

\begin{verbatim}
yabby> seq_load cad3.fas cad3_second

 Reading the file 'cad3.fas' ..
 3 sequence(s) found.

yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    cad3_second    seq           
\end{verbatim}

The two objects 'cad3.seq' and 'cad3\_second.seq' are identical.

\index{print}

The command 'print' allows one to output the sequence object:

\begin{verbatim}
yabby> print cad3.seq

>Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>P95773_STALU [ P95773_STALU ]
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

Upon reading the sequences Yabby has taken the first string
in the FASTA comment to be the sequence ID, and the full
comment is re-inserted within the square brackets. In this
case the only comment was the ID string, and this is merely
repeated within the square brackets. 

The command 'print' can send sequences to a file, instead
of printing them in the terminal window:

\begin{verbatim}
yabby> print -f tmp.fasta cad3.seq

 'cad3.seq' written to the file 'tmp.fasta'
\end{verbatim}

Currently there are several useful options of the 'print'
command. The option '-l' causes protein sequences to be
printed in a three-letter format:

\begin{verbatim}
yabby> print -l cad3.seq

>Q53650_STAAU
TYR VAL ALA THR GLY ILE ASP TYR LEU VAL ILE LEU 
ILE LEU LEU PHE SER GLN VAL LYS LYS GLY GLN VAL 
....further output deleted....
\end{verbatim}

The option '-t N' truncates all sequences at N residues: 

\begin{verbatim}
yabby> print -t 10 cad3.seq

>Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLV
>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLI
>P95773_STALU [ P95773_STALU ]
YIAQALDLLV
\end{verbatim}

\index{seq\_info}

The command 'seq\_info' prints additional information about
sequence objects:

\begin{verbatim}
yabby> seq_info cad3

 'cad3' contains 3 sequence(s)
   min number of residues: 192 (sequence 'Q53650_STAAU')
   max number of residues: 193 (sequence 'Q97PJ0_STRPN')
\end{verbatim}

The option '-l' causes the number of residues to be printed
for each sequence: 

\begin{verbatim}
yabby> seq_info -l cad3

 'cad3' contains 3 sequence(s)
  1 -> Q53650_STAAU, 192 residues
  2 -> Q97PJ0_STRPN, 193 residues
  3 -> P95773_STALU, 192 residues
\end{verbatim}

It is often required to select one or more sequences from the sequence
object. The command 'seq\_pick' allows one to select sequences from
their order number or sequence ID. For example:

\index{seq\_pick}

\begin{verbatim}
yabby> seq_pick -n 2 cad3 s2       

 Fetching the sequence 2 ('Q97PJ0_STRPN')
 Saving the extracted sequence as 's2'
\end{verbatim}

The above command has picked the sequence number 2 from the 'cad3'
object, and saved this sequence under the name 's2': 

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    s2             seq           

yabby> print s2.seq

>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
\end{verbatim}

Another useful option is to select a sequence by its ID string:

\begin{verbatim}
yabby> seq_pick -q Q53650_STAAU cad3 s1

 Fetching the sequence 'Q53650_STAAU'
 Saving the extracted sequence as 's1'

yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    s1             seq           
    s2             seq           
\end{verbatim}

Suppose that we wanted to take the sequence Q53650\_STAAU, and
extract residues 21-40. The above command would take care of the
first part, while the command 'seq\_strip' could be used to
select a residue range:

\index{seq\_strip}

\begin{verbatim}
yabby> seq_strip 21:40 s1 s1_portion

 's1' contains 1 sequence(s)
 stripping 'Q53650_STAAU'

yabby> print s1_portion.seq

>Q53650_STAAU_21:40 [ Q53650_STAAU ]
KGQVKHIWIGQYIGTAIVIG
\end{verbatim}

The command 'seq\_pattern' allows one to search for a pattern
in a sequence. For example, to search for a pattern 'IDY' use: 

\index{seq\_pattern}

\begin{verbatim}
yabby> seq_pattern IDY cad3

 'IDY' matches in 'Q53650_STAAU'
 'IDY' matches in 'Q97PJ0_STRPN'
 3 sequences examined, 2 match(es) found
\end{verbatim} 

The option '-s NAME' allows one to save the matching sequences
under a the new name:

\begin{verbatim}
yabby> seq_pattern -s IDY_matches IDY cad3

 'IDY' matches in 'Q53650_STAAU'
 'IDY' matches in 'Q97PJ0_STRPN'
 3 sequences examined, 2 match(es) found
 Saving matches as 'IDY_matches'
 [ seq_pattern: 'IDY_matches.seq' exists, overwritten ]
\end{verbatim}

The option '-c' allows one to search the sequence comment for a
pattern, rather than the sequence residues:

\begin{verbatim}
yabby> seq_pattern -c STAA cad3

 'STAA' matches in 'Q53650_STAAU'
 3 sequences examined, 1 match(es) found
\end{verbatim}

The command 'seq\_op' allows one to combine two sequence objects
based on their IDs. This command has the following form:

\index{seq\_op}

\begin{verbatim}
seq_op [ options ] SEQ1_OBJ SEQ2_OBJ RES_OBJ
\end{verbatim}

where SEQ1\_OBJ SEQ2\_OBJ are two sequence objects to be combined 
and RES\_OBJ is the name under which the result will be stored.
This command can calculate the union, intersection, symmetric
difference.

For example, consider the sets of sequences stored in files
'cad.fas' and 'cad3.fas', where the first file contains six 
CAD sequences, and cad3.fas contains three out of six sequences
present in 'cad.fas'.

\begin{verbatim}
yabby> seq_load cad.fas cad

 Reading the file 'cad.fas' ..
 6 sequence(s) found.

yabby> seq_load cad3.fas cad3

 Reading the file 'cad3.fas' ..
 3 sequence(s) found.

yabby> seq_op -i cad cad3 cad_i

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 3 sequence(s)
  DIFFERENCE contains 3 sequence(s)
  UNION contains 6 sequence(s)
 [ Saving INTERSECTION as 'cad_i' ]
 [ seq_op: 'cad_i.seq' exists, overwritten ]
\end{verbatim}

The last command has saved the intersection of objects 'cad.seq' and
'cad3.seq' as a new sequence object, named 'cad\_i'. Since 'cad3.seq'
is simply the subset of 'cad.seq' this intersection identical to
'cad3.seq'. Options '-u' and '-d' save the union and difference.
It should be noted that the command 'seq\_op' compares only
sequence IDs, {\em it does not compare the sequences themselves}.
Furthermore, unpredictable results may occur if the sequence IDs
are not unique within each set of sequences. 

\index{seq\_comment}

The command 'seq\_comment' modifies the comment of each sequence
to append the sequence number to its ID. To modify the IDs of the
sequences in 'cad3':

\begin{verbatim}
yabby> seq_comment cad3

 Comments modified in 3 sequence(s).

yabby> print cad3.seq

>1-Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>2-Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>3-P95773_STALU [ P95773_STALU ]
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

Since this command will change the sequence IDs, the output
of the command 'seq\_op' would be different. Specifically,
the intersection of the two sets of sequences would contain
no sequences:

\begin{verbatim}
yabby> seq_op -i cad cad3 cad_u

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 0 sequence(s)
  DIFFERENCE contains 9 sequence(s)
  UNION contains 9 sequence(s)
 [ No sequences to save ]

yabby> seq_op -u cad cad3 cad_u

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 0 sequence(s)
  DIFFERENCE contains 9 sequence(s)
  UNION contains 9 sequence(s)
 [ Saving UNION as 'cad_u' ]
\end{verbatim}

\index{seq\_unique}

The command 'seq\_unique' finds sequences present in one set
not present in the other, by {\em comparing sequence strings}.

\begin{verbatim}
yabby> seq_unique cad cad3 cadu

 3 unique sequences found.
 Saving sequences as 'cadu'
\end{verbatim}

Note that in this case the order of sequence objects is important.
This command find the sequences present in the first sequence
object and not present in the second, i.e.:

\begin{verbatim}
yabby> seq_unique cad3 cad tmp

 No unique sequences found.
\end{verbatim}

\index{seq\_compl}

The command 'seq\_compl' calculates the reverse complement of
a DNA sequence:

\begin{verbatim}
yabby> seq_load dna.fas dna

 Reading the file 'dna.fas' ..
 1 sequence(s) found.

yabby> print dna.seq

>chr01 [ chr01 ]
taaccctaaccctaaccctgaccctaaccctaaccctaaccctaaccctaaccagtacac
gcgtacacgtacaagcacccgtacccccagtatacttggacacccgtactcagttatcct
ttttattagtgtacccgcctcttgcacgcatgccacagttcttcagcagaagaacacgca
caatgctctttgataaacgtgcggacatgaaaaaaagggaaaaacgcagctacgtgtgct
gtcgttggtttcacagcgtcaagccgcgtcggtgtaccaaagaggaggtgacccatcgag

yabby> seq_compl dna dna_c

 'dna' contains 1 sequence(s)
 Working on 'chr01'

yabby> print dna_c.seq

>chr01 [ chr01 ]
attgggattgggattgggactgggattgggattgggattgggattgggattggtcatgtg
cgcatgtgcatgttcgtgggcatgggggtcatatgaacctgtgggcatgagtcaatagga
aaaataatcacatgggcggagaacgtgcgtacggtgtcaagaagtcgtcttcttgtgcgt
gttacgagaaactatttgcacgcctgtactttttttccctttttgcgtcgatgcacacga
cagcaaccaaagtgtcgcagttcggcgcagccacatggtttctcctccactgggtagctc
\end{verbatim}

The option '-r' calculates the reverse complement: 

\begin{verbatim}
yabby> seq_compl -r dna dna_cr

 'dna' contains 1 sequence(s)
 Working on 'chr01'

yabby> print dna_cr.seq

>chr01 [ chr01 ]
ctcgatgggtcacctcctctttggtacaccgacgcggcttgacgctgtgaaaccaacgac
agcacacgtagctgcgtttttccctttttttcatgtccgcacgtttatcaaagagcattg
tgcgtgttcttctgctgaagaactgtggcatgcgtgcaagaggcgggtacactaataaaa
aggataactgagtacgggtgtccaagtatactgggggtacgggtgcttgtacgtgtacgc
gtgtactggttagggttagggttagggttagggttagggtcagggttagggttagggtta
\end{verbatim}

\index{seq\_fetch}

The command 'seq\_fetch' fetches a sequence from GenBank
by the sequence accession number:

\begin{verbatim}
yabby> seq_fetch J00522 mig

 Saving 'J00522' as 'mig'

yabby> what

    object(s)      type
  ------------------------------
    mig            seq           
\end{verbatim}

This command requires the module Bio::DB::GenBank from
BioPerl \cite{bioperl}, and is an example of how easy is
to leverage the capabilities of other bioinformatics packages
within Yabby.

\section{Housekeeping commands}

\index{help}

Several commands are dedicated to the general manipulation of
Yabby objects. Those include 'what' and 'print', and also
'help, 'dump', 'restore', 'delete', 'flush', 'license', 'exit'
and 'quit'.

The 'help' command is probably the most useful. By itself it
prints the list of currently available commands: 

\begin{verbatim}
yabby> help

 31 commands ready.

 blast          blast_info     blastg         delete        
 dump           emboss_needle  exit           flush         
 help           hmm_score      hmm_score2seq  license       
 motif_cmp      motif_load     motif_meme     pfam_fetch    
 print          quit           restore        seq_comment   
 seq_compl      seq_fetch      seq_info       seq_load      
 seq_op         seq_pattern    seq_pick       seq_strip     
 seq_unique     sprot_split    what          

 For info about a particular command try: 'help COMMAND'
\end{verbatim}

If followed by a command name, it prints the information
about the command:

\begin{verbatim}
yabby> help seq_load


 Loads sequence(s) from the database file.

 Usage:
        seq_load [ options ] DBA_FILE OBJ_NAME

 Where DBA_FILE is the name of the database file. OBJ_NAME is
 the internal Yabby name for the sequence(s).

 Options:

 -a -- append sequences to an already existing sequence
  object OBJ_NAME
 -f -- the file format is FASTA (default)
 -b -- the file format is BLOCKS

 Notes:
 1. Only BLOCKS format as given by MEME output was tested.
\end{verbatim}

\index{delete}
\index{flush}

The command 'delete' deletes an object, and the command 'flush'
deletes all objects from the workspace:

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    hits           seq           
                   hmm_score     

yabby> delete hits.seq

 [ 'hits.seq' deleted ]

yabby> flush

 [ workspace flushed ]

yabby> what

    object(s)      type
  ------------------------------
   [ none ]
\end{verbatim}

\index{dump}
\index{restore}

The command 'dump' stores the session onto a file:

\begin{verbatim}
yabby> dump mysession

 Yabby session archived as 'mysession.tar.gz'
\end{verbatim}

The session can be restored with the command 'restore':

\begin{verbatim}
yabby> restore mysession

 Yabby session 'mysession' restored
\end{verbatim}

Note that the commands 'dump'/'restore' depend on unix commands
'tar' and 'gzip' being available for execution by Perl.

\section{Working with sequence motifs}

\index{sequence motifs}

Sequence motifs are handled as 'motif' objects.  Motifs could
be loaded from FASTA files, or from BLOCKS files:

\index{motif\_load}

\begin{verbatim}
yabby> motif_load -b m2.blocks m2

 Reading the file 'm2.blocks' ..
 11 sequence(s) found in the motif 'm2'.
\end{verbatim}

Motifs can be printed:

\begin{verbatim}
yabby> what      

    object(s)      type
  ------------------------------
    m2             motif         

yabby> print m2.motif

>Sthermophile [ from BLOCKS file 'm2.blocks' ]
PMAESGVTIRSDSEQYSRHEEQSVSPPSSSS
>Panserina [ from BLOCKS file 'm2.blocks' ]
PMAESGVTIRSDSEQYSSPEELSTSPPSSSS
>Cglobosum [ from BLOCKS file 'm2.blocks' ]
PMAESGITIPSDSEQYSRHGDQSTSPPSSSS
>Ncrassa [ from BLOCKS file 'm2.blocks' ]
PLAESGVTISSDSEQYSAPESASPQSPSSSS
>Oclavigerum [ from BLOCKS file 'm2.blocks' ]
PMAESGVTMHSDSEQYSGAEELSASLESSHS
>Afumigatus [ from BLOCKS file 'm2.blocks' ]
ELYDSGLTVPSDSEIYSANHEVSSPMSASSS
>Gzea [ from BLOCKS file 'm2.blocks' ]
HLAESGVTMHSDIELYSAGDDLSSPPSSNSP
>Aoryzae [ from BLOCKS file 'm2.blocks' ]
ELYESGLTVRSDSENYSANNELSESTSSSPL
>Tlanuginosus [ from BLOCKS file 'm2.blocks' ]
ELSDSALTVPSDSENYSANNEFSSSPSASNS
>Pnodorum [ from BLOCKS file 'm2.blocks' ]
PLTASGLTIPTDSESYSAPADSPSPSPPSSS
>Apullulans [ from BLOCKS file 'm2.blocks' ]
RDIYDSMTMQSDSETYDQPDISSPSSPSSDS
\end{verbatim}

\index{motif\_cmp}

The command 'motif\_cmp' compares two motifs. Two motifs are
identical if they have the same number of sequences, the sequences
have the same ID, and the sequences themselves are identical
as strings:

\begin{verbatim}
yabby> motif_cmp m2 m2_second

 Motifs 'm2' and 'm2_second' contain the same sequence IDs.
 Comparing the sequences...
 Motifs 'm2' and 'm2_second' are identical.
\end{verbatim}

\index{motif\_meme}
\index{MEME}

The command 'motif\_meme' extracts the motifs from the plain
text output of the motif detection program MEME \cite{meme}.
To extract the motif 1 from the file 'meme.out' use:

\begin{verbatim}
yabby> motif_meme meme.out 1 m1 

 Reading MEME output 'meme.out' ..
 Motif 1 saved as 'm1'.

yabby> what

    object(s)      type
  ------------------------------
    m1             motif         
\end{verbatim}

In the internal representation 'motif' type object is identical to
the 'seq' type object.

\section{Working with the HMMER output}

\index{HMMER}
\index{HMMPFAM}
\index{hmm\_score}
\index{hmm\_score2seq}

HMMER is a powerful software for the profile hidden Markov model
search of sequence databases \cite{hmmer,wwwhmmer}. The 'search 
step' in HMMER search pipeline is performed by the program
HMMPFAM which is a part of the HMMER software suite. Yabby's
commands 'hmm\_score' and 'hmm\_score2seq' can aid in extraction
of the search results from the HMMPFAM output files.

Consider a mock-up example of some the HMMPFAM search performed
on the genome sequence 'LmjFmockup.pep'. The output of this search
was redirected to the file 'hmmpfam.out'. To extract all the hits
with the E-value above 0.01, and store these hits under the name
'hits':

\begin{verbatim}
yabby> hmm_score -E 0.01 -s hits hmmpfam.out

 Processing HMMPFAM search output file

  No     Sequence                            E-score
 -----------------------------------------------------
 (   1)  LmjF05.1190@All                     2.70e-03
 (   2)  LmjF05.1190@GlycogenStarch          4.20e-03
 (   3)  LmjF05.0920@All                     7.50e-03
yabby> what

    object(s)      type
  ------------------------------
    hits           hmm_score     
\end{verbatim}

The command 'hmm\_score' with the '-s hits' flag has created
the object of the type 'hmm\_score' in the workspace. This is
merely a list containing the sequence name, the model the
hit was recorded to, and the E-value:

\begin{verbatim}
LmjF05.1190@All 0.0027
LmjF05.1190@GlycogenStarch 0.0042
LmjF05.0920@All 0.0075
\end{verbatim}

To get to the actual sequences, one needs to go to the
original sequence database on which the HMMPFAM search
was executed, match the sequence IDs, and extract the
sequences. This can be achieved with the command
'hmm\_score2seq'

\begin{verbatim}
yabby> hmm_score2seq LmjFmockup.pep hits.fas hits

 found 3 sequences to extract
 Processing the database 'LmjFmockup.pep'
 Sequences written to 'hits.fas'
\end{verbatim}

This command takes the name of the sequence database on
which the search was executed (LmjFmockup.pep), the name
of the output file where sequence are to be saved (hits.fas),
and the name of the Yabby 'hmm\_score' object where the
hits were extracted.

The output file contains sequence hits in the FASTA format,
which can be loaded and manipulated with sequence commands: 

\begin{verbatim}
yabby> seq_load hits.fas  hits

 Reading the file 'hits.fas' ..
 3 sequence(s) found.

yabby> print hits.seq

>LmjF05.1190@All [ LmjF05.1190@All [ (score: 2.70e-03, model: All) LmjF05.1190 LmjF05.1190 hypothetical protein, conserved 1912350:1918232 reverse MW:212991 ] ]
MSNKKQTWANAHSQLWTSSVARSAGSRTQSEVSSIASTNRDRSLLDDGDGYHQPPSHVCA
QELRHQAYQPNDKTLAQVSDILESYGAIACDADIPAVLLEVVKELERTKRDNNFKDLLLR
EYSDTVQRRFGLYGEESPPSIAEVSARLRDGAPARPVAAPNPAWLEQPLNELWRTVCDGM
QACFEKNADLSAPVLLPEARRTKANVSQLLHTSCDVLSQLAKEYTTAKAAVMKKMERTVA
ASQSFRQLTLSTALSELEAQQELKGSAAEDGGASAMDGSHTVGSALQSLIDVIPASLQIH
.... extra output deleted ....
\end{verbatim}


\section{Running NCBI blast from Yabby}

BLAST (Basic Local Alignment Search Tool) refers to a heuristic
algorithm widely used to find regions of similarity between
protein/DNA sequences. BLAST is often used to denote the computer
program, an actual implementation of the BLAST algorithm. Two
such implementations are in wide use: NCBI BLAST and WU-BLAST
(from the Washington University). In addition, the term BLAST
is often used to denote the Web interface for blast search,
such as the one provided by NCBI \cite{wwwblast}.

It is generally less known that NCBI blast executables can be
downloaded \cite{ftpblast} and run locally, which has some
distinct advantages. The Yabby command 'blast' allows one to
run NCBI blast search locally, through the Yabby interface. 

The command 'blast' depends on NCBI blast being installed locally.
The location of the NCBI blast installation is set in the
Yabby library 'yabby\_blast.pm':

\begin{verbatim}
$BLASTALL = "/usr/local/bin/blastall";
\end{verbatim}

For NCBI BLAST to work one also needs to set up the file
'.ncbirc' in the user's home directory that will give the
location of blast data files (such as substitution matrices): 

\begin{verbatim}
[NCBI]
   Data=/usr/local/blast/2.2.9/data
\end{verbatim}

The database to be searched need first to be indexed for BLAST
search with the program 'formatdb' which is provided in the
NCBI BLAST package.  For example, to format the database
'LmjFwholegenome.pep' (L.major genome in FASTA fomat),

\begin{verbatim}
$ /usr/local/bin/formatdb -i LmjFwholegenome.pep -p T -o T
\end{verbatim}

This would create several additional files in the same directory,
such as 'LmjFwholegenome.pep.phr', 'LmjFwholegenome.pep.psq' etc.

Yabby can execute BLAST search against multiple sequences. For
example, to BLAST six sequences from the file 'cad.fas':

\begin{verbatim}
yabby> seq_load cad.fas cad

 Reading the file 'cad.fas' ..
 6 sequence(s) found.

yabby> blast -E 5.0 tmp/LmjFwholegenome_20050601_V5.2.pep cad

 6 sequence(s) found in the object 'cad'

 Now running BLAST ..
 BLASTing sequence 1 of 6 (Q45153_BACFI)
 Query sequence 'Q45153_BACFI'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 Found 2 hits above the threshold (E=5.00)
 The best hit: 'LmjF36.2210'
 E-score = 1.93e+00

 BLASTing sequence 2 of 6 (Q53650_STAAU)
 Query sequence 'Q53650_STAAU'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 Found 1 hits above the threshold (E=5.00)
 The best hit: 'LmjF04.0510'
 E-score = 3.57e+00

 BLASTing sequence 3 of 6 (Q97PJ0_STRPN)
 Query sequence 'Q97PJ0_STRPN'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 Found 1 hits above the threshold (E=5.00)
 The best hit: 'LmjF13.1210'
 E-score = 2.11e+00

 BLASTing sequence 4 of 6 (P95773_STALU)
 Query sequence 'P95773_STALU'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 No BLAST hits above the threshold (E=5.00) found.

 BLASTing sequence 5 of 6 (Q9JXN6_NEIMB)
 Query sequence 'Q9JXN6_NEIMB'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 No BLAST hits above the threshold (E=5.00) found.

 BLASTing sequence 6 of 6 (Q9CE92_LACLA)
 Query sequence 'Q9CE92_LACLA'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 No BLAST hits above the threshold (E=5.00) found.
\end{verbatim}

The BLAST search was executed for each sequences in the 'cad' set
in turn, creating six 'blast' objects: 

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    cad            seq           
    cad_1          blast         
    cad_2          blast         
    cad_3          blast         
    cad_4          blast         
    cad_5          blast         
    cad_6          blast         
\end{verbatim}

These can be examined later with the command 'blast\_info':

\begin{verbatim}
yabby> blast_info cad_1

 Query sequence 'Q45153_BACFI'
 Database 'tmp/LmjFwholegenome_20050601_V5.2.pep'
 Found 2 hits above the threshold (E=5.00)
 The best hit: 'LmjF36.2210'
 E-score = 1.93e+00
\end{verbatim}

\section{Stand-alone commands}

A stand-alone commands do not interact with other commands. They
are in effect independent program scripts.  Nevertheless it is
often useful to collect such independent scripts under Yabby
to simplify their maintenance and ensure long-term retention.
An example of such command is 'emboss\_needle'.

\index{emboss}
\index{emboss\_needle}

The command 'emboss\_needle' shows the sequences from the output
of the program 'needle' from the EMBOSS suite of programs
\cite{emboss}. The input to the program 'needle' are two
sequences, and the program calculates the global sequence
alignment (Needleman-Wunsch algorithm). Typically two files
are given on the input, each containing one sequence in FASTA  
format. If the second file contains multiple sequences,
the global sequence alignment will be calculate between
the first sequence and each sequence in turn from the
second set. If the second set contains many sequences, the
output file will be large, and it may be difficult to see
the best matches. These can be extracted with the command
'emboss\_needle':

\begin{verbatim}
yabby> emboss_needle needle.out 3

 Processing the file 'needle.out' ..

 (1) Q45153_BACFI:LmjF05.1170, Similarity: 26.6
 (2) Q45153_BACFI:LmjF05.1120, Similarity: 16.4
 (3) Q45153_BACFI:LmjF05.1040, Similarity: 15.0
\end{verbatim}

The second argument is the number of top-scoring sequences
to be printed out.

\index{pdb\_model.pl}

Another example of a stand-alone command is 'pdb\_model', which 
splits the Protein Data Bank (PDB) file with multiple models.
The output is a series of PDB file, one per model:

\begin{verbatim}
yabby> pdb_model 2C7H.pdb RBP_

 working on model  1 (creating 'RBP_1.pdb')
 working on model  2 (creating 'RBP_2.pdb')
 working on model  3 (creating 'RBP_3.pdb')
 working on model  4 (creating 'RBP_4.pdb')
 working on model  5 (creating 'RBP_5.pdb')
.... extra output deleted ....
\end{verbatim}

where '2C7H.pdb' is the original PDB file with multiple models,
and 'RBP\_' is the prefix used for individual model files.

