% chapter02.tex

 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %                                                                           %
 %    YABBY documentation                                                    %
 %    Copyright (C) 2007 Vladimir Likic                                      %
 %                                                                           %
 %    The files in this directory provided under the Creative Commons        %
 %    Attribution-NonCommercial-NoDerivs 2.1 Australia license               %
 %    http://creativecommons.org/licenses/by-nc-nd/2.1/au/                   %
 %    See the file license.txt                                               %
 %                                                                           %
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\setcounter{section}{0}

\chapter{Tutorial}

The data files used in this tutorial can be found in docs/data/ 
directory. Starting Yabby in this directory should allow one to
execute all examples given in the tutorial.

\section{Working with sequences}

\index{sequences}

In Yabby sequences are represented as sequence objects. A sequence
object may contain one or more sequences. 

One way to create a sequence object is to load sequences from a
file.  Consider the file 'cad3.seq' which contains three sequences
from the Pfam CAD family in the FASTA format:

\begin{verbatim}
>Q53650_STAAU
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>Q97PJ0_STRPN
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>P95773_STALU
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

\index{seq\_load}

To load this file in the workspace use the command 'seq\_load':

\begin{verbatim}
yabby> seq_load cad3.seq cad3

 Reading the file 'cad3.seq' ..
 3 sequence(s) found.
\end{verbatim}

Most Yabby commands which create a new object in the workspace
conform to the same pattern:

\begin{verbatim}
COMMAND [ options ] ARGS OBJ_NAME
\end{verbatim}

Where '[ options ]' is where flags and options are passed (if
any), ARGS are the arguments to the command, and OBJ\_NAME is
the name of the object to be created. In the example given
above, there were no options to the 'seq\_load' command; there
was one argument, the name of the file; and the object to be
created was named 'cad3'.

It is possible to inspect objects currently available in the
workspace:

\index{what}

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
\end{verbatim}

This listing shows that one object is currently in the workspace,
of the 'seq' type. It is possible to load the same sequences
under a different name: 

\begin{verbatim}
yabby> seq_load cad3.seq cad3_second

 Reading the file 'cad3.seq' ..
 3 sequence(s) found.

yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    cad3_second    seq           
\end{verbatim}

The two objects 'cad3.seq' and 'cad3\_second.seq' are identical.

\index{print}

The command 'print' allows one to output the sequence object:

\begin{verbatim}
yabby> print cad3.seq

>Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>P95773_STALU [ P95773_STALU ]
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

Upon reading the sequences Yabby has taken the first string
in the FASTA comment to be the sequence ID, and the full
comment is re-inserted within the square brackets. In this
case the only comment was the ID string, and this is merely
repeated within the square brackets. 

The command 'print' can send sequences to a file, instead
of printing them in the terminal window:

\begin{verbatim}
yabby> print -f tmp.fasta cad3.seq

 'cad3.seq' written to the file 'tmp.fasta'
\end{verbatim}

Currently there are several useful options of the 'print'
command. The option '-l' causes protein sequences to be
printed in a three-letter format:

\begin{verbatim}
yabby> print -l cad3.seq

>Q53650_STAAU
TYR VAL ALA THR GLY ILE ASP TYR LEU VAL ILE LEU 
ILE LEU LEU PHE SER GLN VAL LYS LYS GLY GLN VAL 
....further output deleted....
\end{verbatim}

The option '-t N' truncates all sequences at N residues: 

\begin{verbatim}
yabby> print -t 10 cad3.seq

>Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLV
>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLI
>P95773_STALU [ P95773_STALU ]
YIAQALDLLV
\end{verbatim}

\index{seq\_info}

The command 'seq\_info' prints additional information about
sequence objects:

\begin{verbatim}
yabby> seq_info cad3

 'cad3' contains 3 sequence(s)
   min number of residues: 192 (sequence 'Q53650_STAAU')
   max number of residues: 193 (sequence 'Q97PJ0_STRPN')
\end{verbatim}

The option '-l' causes the number of residues to be printed
for each sequence: 

\begin{verbatim}
yabby> seq_info -l cad3

 'cad3' contains 3 sequence(s)
  1 -> Q53650_STAAU, 192 residues
  2 -> Q97PJ0_STRPN, 193 residues
  3 -> P95773_STALU, 192 residues
\end{verbatim}

It is often required to select one or more sequences from the sequence
object. The command 'seq\_pick' allows one to select sequences from
their order number or sequence ID. For example:

\index{seq\_pick}

\begin{verbatim}
yabby> seq_pick -n 2 cad3 s2       

 Fetching the sequence 2 ('Q97PJ0_STRPN')
 Saving the extracted sequence as 's2'
\end{verbatim}

The above command has picked the sequence number 2 from the 'cad3'
object, and saved this sequence under the name 's2': 

\begin{verbatim}
yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    s2             seq           

yabby> print s2.seq

>Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
\end{verbatim}

Another useful option is to select a sequence by its ID string:

\begin{verbatim}
yabby> seq_pick -q Q53650_STAAU cad3 s1

 Fetching the sequence 'Q53650_STAAU'
 Saving the extracted sequence as 's1'

yabby> what

    object(s)      type
  ------------------------------
    cad3           seq           
    s1             seq           
    s2             seq           
\end{verbatim}

Suppose that we wanted to take the sequence Q53650\_STAAU, and
extract residues 21-40. The above command would take care of the
first part, while the command 'seq\_strip' could be used to
select a residue range:

\index{seq\_strip}

\begin{verbatim}
yabby> seq_strip 21:40 s1 s1_portion

 's1' contains 1 sequence(s)
 stripping 'Q53650_STAAU'

yabby> print s1_portion.seq

>Q53650_STAAU_21:40 [ Q53650_STAAU ]
KGQVKHIWIGQYIGTAIVIG
\end{verbatim}

The command 'seq\_pattern' allows one to search for a pattern
in a sequence. For example, to search for a pattern 'IDY' use: 

\index{seq\_pattern}

\begin{verbatim}
yabby> seq_pattern IDY cad3

 'IDY' matches in 'Q53650_STAAU'
 'IDY' matches in 'Q97PJ0_STRPN'
 3 sequences examined, 2 match(es) found
\end{verbatim} 

The option '-s NAME' allows one to save the matching sequences
under a the new name:

\begin{verbatim}
yabby> seq_pattern -s IDY_matches IDY cad3

 'IDY' matches in 'Q53650_STAAU'
 'IDY' matches in 'Q97PJ0_STRPN'
 3 sequences examined, 2 match(es) found
 Saving matches as 'IDY_matches'
 [ seq_pattern: 'IDY_matches.seq' exists, overwritten ]
\end{verbatim}

The option '-c' allows one to search the sequence comment for a
pattern, rather than the sequence residues:

\begin{verbatim}
yabby> seq_pattern -c STAA cad3

 'STAA' matches in 'Q53650_STAAU'
 3 sequences examined, 1 match(es) found
\end{verbatim}

The command 'seq\_op' allows one to combine two sequence objects
based on their IDs. This command has the following form:

\index{seq\_op}

\begin{verbatim}
seq_op [ options ] SEQ1_OBJ SEQ2_OBJ RES_OBJ
\end{verbatim}

where SEQ1\_OBJ SEQ2\_OBJ are two sequence objects to be combined 
and RES\_OBJ is the name under which the result will be stored.
This command can calculate the union, intersection, symmetric
difference.

For example, consider the sets of sequences stored in files
'cad.seq' and 'cad3.seq', where the first file contains six 
CAD sequences, and cad3.seq contains three out of six sequences
present in 'cad.seq'.

\begin{verbatim}
yabby> seq_load cad.seq cad

 Reading the file 'cad.seq' ..
 6 sequence(s) found.

yabby> seq_load cad3.seq cad3

 Reading the file 'cad3.seq' ..
 3 sequence(s) found.

yabby> seq_op -i cad cad3 cad_i

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 3 sequence(s)
  DIFFERENCE contains 3 sequence(s)
  UNION contains 6 sequence(s)
 [ Saving INTERSECTION as 'cad_i' ]
 [ seq_op: 'cad_i.seq' exists, overwritten ]
\end{verbatim}

The last command has saved the intersection of objects 'cad.seq' and
'cad3.seq' as a new sequence object, named 'cad\_i'. Since 'cad3.seq'
is simply a subset of 'cad.seq' this intersection identical to
'cad3.seq'. Options '-u' and '-d' save the union and difference.
It should be noted that the command 'seq\_op' compares only
sequence IDs, {\em it does not compare the sequences themselves}.
Furthermore, unpredictable results may occur if the sequence IDs
are not unique within each set of sequences. 

\index{seq\_comment}

The command 'seq\_comment' modifies the comment of each sequence
to apped the sequence number to its ID. To modify the IDs of the
sequences in 'cad3':

\begin{verbatim}
yabby> seq_comment cad3

 Comments modified in 3 sequence(s).

yabby> print cad3.seq

>1-Q53650_STAAU [ Q53650_STAAU ]
YVATGIDYLVILILLFSQVKKGQVKHIWIGQYIGTAIVIGASLLVAQGVVNLIPQQWVIG
LLGLLPLYLGVKIWIKGEEDEDESSILSLFSSGKFNQLFLTMIFIVLASSADDFSIYIPY
FTTLSMSEIFIVTIVFLIMVGVLCYVSYRLASFDFISETIEKYERWIVPIVFIGLGIYIL
FENGTSNALISF
>2-Q97PJ0_STRPN [ Q97PJ0_STRPN ]
YISTSIDYLIILIILFAQLSQNKQKWHIYAGQYLGTGLLVGASLVAAYVVNFVPEEWMVG
LLGLIPIYLGIRFAIVGEDAEEEEEEIIERLEQSKANQLFWTVTLLTIASGGDNLGIYIP
YFASLDWSQTLVALLVFVIGIIIFCEISRVLSSIPLIFETIEKYERIIVPLVFILLGLYI
MYENGTIETFLIV
>3-P95773_STALU [ P95773_STALU ]
YIAQALDLLVILLMFFARAKTRKEYRDIYIGQYVGSVALIVISLFFAFVLNYVPEKWILG
LLGLIPIYLGIKVAIYGDSDGEERAKKELNEKGLSKLVGTIAIVTIASCGADNIGLFVPY
FVTLSVTNLLITLFVFLILIFFLVFAAQKLANIPEVGEIVEKFGRWIMAVIYIALGLFII
IENDTIQTILGF
\end{verbatim}

Since this command will change the sequence IDs, the output
of the command 'seq\_op' would be different. Specifically,
the intersection of the two sets of sequences would contain
no sequences:

\begin{verbatim}
yabby> seq_op -i cad cad3 cad_u

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 0 sequence(s)
  DIFFERENCE contains 9 sequence(s)
  UNION contains 9 sequence(s)
 [ No sequences to save ]

yabby> seq_op -u cad cad3 cad_u

 Found 6 sequence(s) in 'cad'
 Found 3 sequence(s) in 'cad3'
  INTERSECTION contains 0 sequence(s)
  DIFFERENCE contains 9 sequence(s)
  UNION contains 9 sequence(s)
 [ Saving UNION as 'cad_u' ]
\end{verbatim}

\index{seq\_unique}

The command 'seq\_unique' finds sequences present in one set
not present in the other, by {\em comparing sequence strings}.

\begin{verbatim}
yabby> seq_unique cad cad3 cadu

 3 unique sequences found.
 Saving sequences as 'cadu'
\end{verbatim}

Note that in this case the order of sequence objects is important.
This command find the sequences present in the first sequence
object and not present in the second, i.e.:

\begin{verbatim}
yabby> seq_unique cad3 cad tmp

 No unique sequences found.
\end{verbatim}

\index{seq\_compl}

The command 'seq\_compl' calculates the reverse complement of
a DNA sequence:

\begin{verbatim}
yabby> seq_load dna.seq dna

 Reading the file 'dna.seq' ..
 1 sequence(s) found.

yabby> print dna.seq

>chr01 [ chr01 ]
taaccctaaccctaaccctgaccctaaccctaaccctaaccctaaccctaaccagtacac
gcgtacacgtacaagcacccgtacccccagtatacttggacacccgtactcagttatcct
ttttattagtgtacccgcctcttgcacgcatgccacagttcttcagcagaagaacacgca
caatgctctttgataaacgtgcggacatgaaaaaaagggaaaaacgcagctacgtgtgct
gtcgttggtttcacagcgtcaagccgcgtcggtgtaccaaagaggaggtgacccatcgag

yabby> seq_compl dna dna_c

 'dna' contains 1 sequence(s)
 Working on 'chr01'

yabby> print dna_c.seq

>chr01 [ chr01 ]
attgggattgggattgggactgggattgggattgggattgggattgggattggtcatgtg
cgcatgtgcatgttcgtgggcatgggggtcatatgaacctgtgggcatgagtcaatagga
aaaataatcacatgggcggagaacgtgcgtacggtgtcaagaagtcgtcttcttgtgcgt
gttacgagaaactatttgcacgcctgtactttttttccctttttgcgtcgatgcacacga
cagcaaccaaagtgtcgcagttcggcgcagccacatggtttctcctccactgggtagctc
\end{verbatim}

The option '-r' calculates the reverse complement: 

\begin{verbatim}
yabby> seq_compl -r dna dna_cr

 'dna' contains 1 sequence(s)
 Working on 'chr01'

yabby> print dna_cr.seq

>chr01 [ chr01 ]
ctcgatgggtcacctcctctttggtacaccgacgcggcttgacgctgtgaaaccaacgac
agcacacgtagctgcgtttttccctttttttcatgtccgcacgtttatcaaagagcattg
tgcgtgttcttctgctgaagaactgtggcatgcgtgcaagaggcgggtacactaataaaa
aggataactgagtacgggtgtccaagtatactgggggtacgggtgcttgtacgtgtacgc
gtgtactggttagggttagggttagggttagggttagggtcagggttagggttagggtta
\end{verbatim}

\index{seq\_fetch}

The command 'seq\_fetch' fetches a sequence from GenBank
by the sequence accession number:

\begin{verbatim}
yabby> seq_fetch J00522 mig

 Saving 'J00522' as 'mig'

yabby> what

    object(s)      type
  ------------------------------
    mig            seq           
\end{verbatim}

This command requires the module Bio::DB::GenBank from
BioPerl \cite{bioperl}, and is an example of how easy is
to leverage the capabilities of other bioinformatics packages
within Yabby.


\section{Working with sequence motifs}

\index{sequence motifs}

Sequence motifs are handled as 'motif' objects.  Motifs could
be loaded from FASTA files, or from BLOCKS files:



In the internal representation this type object is identical to
the sequence type object.

